utils/rtc_clock_menu.c

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (C) 2022 PC Engines GmbH
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+
+#include <libpayload.h>
+#include <curses.h>
+
+static void print_rtc_clock_menu(struct tm *t) {
+	printf("\n\n--- Clock menu ---\n");
+	printf("  Date: %04d-%02d-%02d\n", t->tm_year + 1900, t->tm_mon + 1, t->tm_mday);
+	printf("  Time: %02d:%02d:%02d\n", t->tm_hour, t->tm_min, t->tm_sec);
+	printf("  Y year   - Set year\n");
+	printf("  M month  - Set month\n");
+	printf("  D day    - Set day\n");
+	printf("  h hour   - Set hour\n");
+	printf("  m minute - Set minute\n");
+	printf("  s second - Set second\n");
+	printf("  w - Write and exit\n");
+	printf("  x - Exit without writing\n");
+	printf("\n");
+}
+
+static void handle_year_command(struct tm *t, char *command) {
+	int year;
+
+	year = strtoul(command+2, NULL, 10);
+	if (year >= 1900 && year <= 2099)
+		t->tm_year = year - 1900;
+	else
+		printf("Please enter a year between 1900 and 2099\n");
+}
+
+static void handle_mon_command(struct tm *t, char *command) {
+	int mon;
+
+	mon = strtoul(command+2, NULL, 10);
+	if (mon >= 1 && mon <= 12)
+		t->tm_mon = mon - 1;
+	else
+		printf("Please enter a month between 1 and 12\n");
+}
+
+static void handle_day_command(struct tm *t, char *command) {
+	int day;
+
+	day = strtoul(command+2, NULL, 10);
+	if (day >= 1 && day <= 31)
```

Comment:
Oooh, can we set 31st of February? Interesting...

Reply:
Fixed.

Reply:
Yes, any OS would fix it and sync the time via NTP or something... But... are we sure any software is capable of that? Think about the issue reported on GitHub.